Example generations using dictionary in both interactive and command-line modes.

StarKhan6368 opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->

**What**: Better Example generations by utilizing dictionary in both interactive and command-line modes.

<!-- Why are these changes necessary? -->

**Why**:  
Currently, examples are generated using random values, but now users can specify a dictionary for value substitution when generating examples.

<!-- How were these changes implemented? -->

**How**:
By leveraging the `substituteDictionaryValues` method in a Dictionary class, we can enable `forceSubstitution`, which replaces placeholder values with actual values from a user-provided dictionary.

<!-- Have you done all of these things?  -->

**Checklist**:

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->

<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

- [ ] Documentation added to the README.md OR link to PR on https://github.com/specmatic/specmatic-documentation
- [x] Tests
- [ ] Sonar Quality Gate

<!-- Kindly link the issue that this PR will be addressing -->

<!-- feel free to add additional comments -->

StarKhan6368 committed:
Add dictionary based substitution support.

- Use dictionary values in example generation.

joelrosario committed:
Renamed vars and functions

joelrosario committed:
Refactored the example validation commands using the same core logic for validation

joelrosario committed:
When showing validation errors, correctly denote the example type

joelrosario committed:
Cleaned up duplicate invocations of the common function that returns a failure

joelrosario committed:
/_specmatic/v2/examples/validate now uses the validation loop from validate. It then loops over the results to transform them into responses.

joelrosario committed:
Fixed output formatting

joelrosario committed:
Fixed regression in which an example with no errors would print an empty example

joelrosario committed:
Ignore inline examples belonging to scenarios that have been excluded using the filter parameters

joelrosario committed:
Fixed issue with URLs in the GUI resulting in a 404 when generating or validating an example for an operation

StarKhan6368 committed:
Refactor, default dictionary in example gen.

- Move dictionary substitution methods to a common
  dictionary class and use the class instead of
  map of string, value
- Look for dictionary.json incase no --dictionary
  provided in example generation.

joelrosario committed:
Added tests for the main functionality of generation and validation

StarKhan6368 committed:
Add simple tests for Dictionary class.

- testing `forceSubstitution` argument.

joelrosario committed:
Updated a test name

joelrosario commented on core/src/main/kotlin/io/specmatic/core/Substitution.kt:
```diff
@@ -106,7 +106,7 @@ class Substitution(
         val name = string.trim().removeSurrounding("$(", ")")
         return variableValues[name]
-                ?: key?.let { dictionary[key]?.toStringLiteral() }
+                ?: key?.let { dictionary.map[key]?.toStringLiteral() }
```
The map property of the dictionary should be wrapped. Nothing in the dictionary should be directly referred to outside. You could for example turn this into: 

`dictionary.lookup(key).toStringLiteral()` with the method signature of `lookup` being `fun lookup(key: String): Value`

joelrosario commented on core/src/test/kotlin/io/specmatic/core/DictionaryTest.kt:
```diff
@@ -0,0 +1,46 @@
+
+// NOTE: Concrete values are actual values, not placeholders like (string) or (number).
+// For example, "ATYGHA" is a concrete value, not a placeholder.
+class DictionaryTest {
```
We need a test somewhere of how generation will work when a dictionary is involved.

joelrosario commented on core/src/main/kotlin/io/specmatic/core/Dictionary.kt:
```diff
@@ -0,0 +1,78 @@
+        }
+    }
+
+    fun substituteDictionaryValues(httpResponse: HttpResponse, forceSubstitution: Boolean = false): HttpResponse {
```
We need to make this functionality generally available without tying it to HTTP requests and responses. Ideally these methods should be in their respective classes.

StarKhan6368 committed:
Dictionary map should be inaccessible outside.

- Add tests for example generation with
  external dictionary involved.
- move HttpRequest and HttpResponse sub function
  (s) to their respective classes.
- provide substitute function for headers
  in Dictionary.kt

joelrosario committed:
Reduced flakiness of proxy test

StarKhan6368 committed:
Merge branch 'refs/heads/example_generation_improvements' into dict_based_example_gen_merged

# Conflicts:
#	application/src/main/kotlin/application/ExamplesCommand.kt
#	core/src/main/kotlin/io/specmatic/core/Feature.kt
#	core/src/main/kotlin/io/specmatic/core/examples/server/ExamplesInteractiveServer.kt

joelrosario committed:
Removed some exitProcess and exitWithMessage function calls as they kill tests early

joelrosario committed:
Small improvement to the example generation error message
